sql/stats: TestCreateStatsUsingExtremesProgress failed

sql/stats.TestCreateStatsUsingExtremesProgress [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17841754?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17841754?buildTab=artifacts#/) on master @ [a7a035075a608ecbfaaf2c565f67efa6cb4bf63f](https://github.com/cockroachdb/cockroach/commits/a7a035075a608ecbfaaf2c565f67efa6cb4bf63f):

Fatal error:

```
panic: test timed out after 5m0s
running tests:
	TestCreateStatsUsingExtremesProgress (4m41s)

```
Stack: 

```
goroutine 43701 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4003a147a8?)
* 	external/com_github_cockroachdb_pebble/table_cache.go:397 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 10572
* 	external/com_github_cockroachdb_pebble/table_cache.go:389 +0xe4
* 
* goroutine 34919 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x7c
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 10572
* 	pkg/util/admission/work_queue.go:421 +0x2ec
* 
* goroutine 42237 [sync.Cond.Wait, 2 minutes]:
* sync.runtime_notifyListWait(0x400da5e0a8, 0x7d)
* 	GOROOT/src/runtime/sema.go:569 +0x154
* sync.(*Cond).Wait(0x400da5e098)
* 	GOROOT/src/sync/cond.go:70 +0xcc
* github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:371
* github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0x400da5e000, {0x4787ae0?, 0x9be8c60?})
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:645 +0x670
* runtime/pprof.Do({0x656b638?, 0x9be8c60?}, {{0x4000e2e0a0?, 0x0?, 0x0?}}, 0x400d7ba7b8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:564 +0x60
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 35298
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:563 +0x44c
* 
* goroutine 34896 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x656b830?, 0x401060dd10?})
* 	external/com_github_cockroachdb_pebble/table_cache.go:399 +0x80
* runtime/pprof.Do({0x656b638?, 0x9be8c60?}, {{0x4000e2e8c0?, 0x4002f018c0?, 0x4000d7d860?}}, 0x4005455f98)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4005455fa8?)
* 	external/com_github_cockroachdb_pebble/table_cache.go:397 +0x5c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 10572
* 	external/com_github_cockroachdb_pebble/table_cache.go:389 +0xe4
* 
* goroutine 34922 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	pkg/util/admission/work_queue.go:501 +0x140
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 10572
* 	pkg/util/admission/work_queue.go:472 +0x5c
* 
* goroutine 34925 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x7c
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 10572
* 	pkg/util/admission/work_queue.go:421 +0x2ec
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCreateStatsUsingExtremesProgress.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44690

--- a/pkg/sql/stats/create_stats_job_test.go
+++ b/pkg/sql/stats/create_stats_job_test.go
@@ -564,14 +564,6 @@ func TestCreateStatsProgress(t *testing.T) {
 	skip.UnderDeadlock(t, "the test is too sensitive to overload")
 
 	var allowRequest chan struct{}
-	var allowRequestClosed bool
-	// Make sure that we unblock the test server in all scenarios with test
-	// failures.
-	defer func() {
-		if !allowRequestClosed {
-			close(allowRequest)
-		}
-	}()
 	filter, setTableID := createStatsRequestFilter(&allowRequest)
 	var params base.TestServerArgs
 	params.Knobs.Store = &kvserver.StoreTestingKnobs{
@@ -588,6 +580,15 @@ func TestCreateStatsProgress(t *testing.T) {
 	defer srv.Stopper().Stop(ctx)
 	s := srv.ApplicationLayer()
 	sqlDB := sqlutils.MakeSQLRunner(conn)
+
+	var allowRequestClosed bool
+	// Make sure that we unblock the test server in all scenarios with test
+	// failures.
+	defer func() {
+		if !allowRequestClosed {
+			close(allowRequest)
+		}
+	}()
 
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false`)
 	sqlDB.Exec(t, `CREATE DATABASE d`)
@@ -723,14 +724,6 @@ func TestCreateStatsUsingExtremesProgress(t *testing.T) {
 	skip.UnderDeadlock(t, "the test is too sensitive to overload")
 
 	var allowRequest chan struct{}
-	var allowRequestClosed bool
-	// Make sure that we unblock the test server in all scenarios with test
-	// failures.
-	defer func() {
-		if !allowRequestClosed {
-			close(allowRequest)
-		}
-	}()
 	filter, setTableID := createStatsRequestFilter(&allowRequest)
 	var params base.TestServerArgs
 	params.Knobs.Store = &kvserver.StoreTestingKnobs{
@@ -747,6 +740,15 @@ func TestCreateStatsUsingExtremesProgress(t *testing.T) {
 	defer srv.Stopper().Stop(ctx)
 	s := srv.ApplicationLayer()
 	sqlDB := sqlutils.MakeSQLRunner(conn)
+
+	var allowRequestClosed bool
+	// Make sure that we unblock the test server in all scenarios with test
+	// failures.
+	defer func() {
+		if !allowRequestClosed {
+			close(allowRequest)
+		}
+	}()
 
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false`)
 	sqlDB.Exec(t, `CREATE DATABASE d`)
@@ -777,7 +779,7 @@ func TestCreateStatsUsingExtremesProgress(t *testing.T) {
 	// Invalidate the stats cache so that we can be sure to get the latest stats.
 	s.ExecutorConfig().(sql.ExecutorConfig).TableStatsCache.InvalidateTableStats(ctx, tID)
 
-	sqlDB.Exec(t, `INSERT INTO d.t SELECT x, x from generate_series(101, 120) as g(x)`)
+	sqlDB.Exec(t, `INSERT INTO d.t SELECT x, x from generate_series(101, 1000) as g(x)`)
 
 	const partialStatQuery = `CREATE STATISTICS s2 FROM d.t USING EXTREMES`
 
@@ -788,13 +790,24 @@ func TestCreateStatsUsingExtremesProgress(t *testing.T) {
 		_, err := conn.Exec(partialStatQuery)
 		errCh <- err
 	}()
-	allowRequest <- struct{}{}
+	// Ten iterations here allows us to read some of the rows but not all.
+	for i := 0; i < 10; i++ {
+		select {
+		case allowRequest <- struct{}{}:
+		case err := <-errCh:
+			if err == nil {
+				t.Fatalf("query unexpectedly finished")
+			} else {
+				t.Fatal(err)
+			}
+		}
+	}
 
 	// Fetch the new job ID since we know it's running now.
 	jobID := getLastCreateStatsJobID(t, sqlDB)
 
 	var fractionCompleted float32
-	var prevFractionCompleted float32
+	prevFractionCompleted := getFractionCompleted(t, sqlDB, jobID)
 
 	// Allow the job to progress until it finishes scanning both indexes.
 Loop:
@@ -885,8 +898,11 @@ func createStatsRequestFilter(
 			_, tableID, _ := encoding.DecodeUvarintAscending(key)
 			// Ensure that the tableID is what we expect it to be.
 			if tableID > 0 && descpb.ID(tableID) == tableToBlock.Load() {
-				// Read from the channel twice to allow jobutils.RunJob to complete
-				// even though there is only one ScanRequest.
+				// Read from the channel twice to allow runCreateStatsJob to
+				// complete even though there is only one ScanRequest.
+				// TODO(yuzefovich): only some tests need this behavior.
+				// Consider asking the caller how many times we should receive
+				// from the channel.
 				<-*allowProgressIota
 				<-*allowProgressIota
 			}

